### PR TITLE
xwayland-run: init at 0.0.2

### PR DESCRIPTION
**Pull request by arthsmn**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/xw/xwayland-run/package.nix [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    ninja
+  ];
+
+  buildInputs = [
+    python3
+  ];
+
+  meta = with lib; {
+    description = "A set of small utilities revolving around running Xwayland
+and various Wayland compositor headless.";
```
Comment (SuperSandro2000) [inline pkgs/by-name/xw/xwayland-run/package.nix:32]:
```suggestion
    description = "A set of small utilities revolving around running Xwayland and various Wayland compositor headless.";
```

**Review by SuperSandro2000 — APPROVED** [review]

File: pkgs/by-name/xw/xwayland-run/package.nix [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    meson
+    ninja
+  ];
+
+  buildInputs = [
+    python3
+  ];
+
+  meta = with lib; {
+    description = "A set of small utilities revolving around running Xwayland and various Wayland compositor headless.";
```
Comment (AndersonTorres) [inline pkgs/by-name/xw/xwayland-run/package.nix:31]:
```suggestion
    description = "A set of small utilities revolving around running Xwayland and various Wayland compositor headless";
```

No ending dot.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/by-name/xw/xwayland-run/package.nix (hunk 2 above)
Comment (arthsmn) [reply pkgs/by-name/xw/xwayland-run/package.nix]: Done.

**Review by AndersonTorres — APPROVED** [review]

**Comment by pbsds** [comment]
It seems the entrypoints need wrapping to properly enter the python environment

```
$ results/xwayland-run/bin/xwfb-run
Traceback (most recent call last):
  File "/dev/shm/nixpkgs-review/pr-273453/results/xwayland-run/bin/xwfb-run", line 30, in <module>
    from wlheadless.wlheadless_common import WlheadlessCommon
ModuleNotFoundError: No module named 'wlheadless'
```

Consider `buildPythonApplication` with `pyproject=false` instead of `stdenv.mkDerivation`

**Comment by arthsmn** [comment]
> It seems the entrypoints need wrapping to properly enter the python environment
> 
> ```
> $ results/xwayland-run/bin/xwfb-run
> Traceback (most recent call last):
>   File "/dev/shm/nixpkgs-review/pr-273453/results/xwayland-run/bin/xwfb-run", line 30, in <module>
>     from wlheadless.wlheadless_common import WlheadlessCommon
> ModuleNotFoundError: No module named 'wlheadless'
> ```
> 
> Consider `buildPythonApplication` with `pyproject=false` instead of `stdenv.mkDerivation`

Sorry! Done.

File: pkgs/by-name/xw/xwayland-run/package.nix [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+, fetchFromGitLab
+, meson
+, ninja
+, python3
+}:
+
+python3.pkgs.buildPythonApplication rec {
+  pname = "xwayland-run";
+  version = "0.0.2";
+
```
Comment (pbsds) [inline pkgs/by-name/xw/xwayland-run/package.nix:11]:
```suggestion

  outputs = [ "out" "man" ];

```

**Review by pbsds — COMMENTED** [review]
On a headless host i get no output and no side-effects when running a program:

```
$ /nix/store/798xf6snhdbyi7z7c8s3ism06mxq8fv2-xwayland-run-0.0.2/bin/xwfb-run touch foobar
$ test -f foobar || echo ":("
:(
```

In `strace` i can see it handling some error:

```
read(4, "# subprocess - Subprocesses with"..., 8192) = 8192
read(4, "PollSelector'):\n        _PopenSe"..., 8192) = 8192
read(4, "elf.returncode = returncode\n    "..., 8192) = 8192
read(4, ".\n\n    Prefer an implementation "..., 8192) = 8192
read(4, "      raise ValueError(\"The 'ext"..., 8192) = 8192
read(4, "subprocess exit status creates a"..., 8192) = 8192
read(4, "f _get_handles(self, stdin, stdo"..., 8192) = 8192
read(4, "ot executable:\n                 "..., 8192) = 8192
```

On a wayland gnome host i get no side-effects and the following error:
```
Failed to connect to the compositor!
```

When i run it in `cage` i get the following output, not sure if relevant:

```
$ cage results/xwayland-run/bin/xwfb-run touch asd
00:00:00.012 [EGL] command: eglQueryDmaBufModifiersEXT, error: EGL_BAD_PARAMETER (0x300c), message: "EGL_BAD_PARAMETER error: In eglQueryDmaBufModifiersEXT: Invalid format
"
00:00:00.012 [render/egl.c:661] Failed to query dmabuf number of modifiers
00:00:00.012 [EGL] command: eglQueryDmaBufModifiersEXT, error: EGL_BAD_PARAMETER (0x300c), message: "EGL_BAD_PARAMETER error: In eglQueryDmaBufModifiersEXT: Invalid format
"
00:00:00.012 [render/egl.c:661] Failed to query dmabuf number of modifiers
Failed to connect to the compositor!
```

File: pkgs/by-name/xw/xwayland-run/package.nix (hunk 3 above)
Comment (arthsmn) [reply pkgs/by-name/xw/xwayland-run/package.nix:11]: Actually when I built with this the man directory from result/share disappeared.

Comment (SuperSandro2000) [reply pkgs/by-name/xw/xwayland-run/package.nix:11]: it is now in the man output which you might need to explicitly choose to build.

Comment (arthsmn) [reply pkgs/by-name/xw/xwayland-run/package.nix:11]: Oh, understood. Done.

**Comment by arthsmn** [comment]
@pbsds everything should be fine now.

**Comment by pbsds** [comment]
Seems to work, but putting both weston, cage and mutter in the closure is quite excessive.
In `share/wlheadless/wlheadless.conf` it seems weston is the default compositor, and cage and mutter are simply configurable options. Lets remove those

Up for discussion is, should we set cage as the default compositor?

Instead of propagating the runtime dependencies, please wrap the `bin/` entrypoints instead.

**Comment by arthsmn** [comment]
> Seems to work, but putting both weston, cage and mutter in the closure is quite excessive. In `share/wlheadless/wlheadless.conf` it seems weston is the default compositor, and cage and mutter are simply configurable options. Lets remove those
>
Yeah agreed.


> Up for discussion is, should we set cage as the default compositor?
>
I think it's better to  keep upstream's defaults.


> Instead of propagating the runtime dependencies, please wrap the `bin/` entrypoints instead.
>
I'll take a look tomorrow. Thanks for the suggestion.

**Comment by arthsmn** [comment]
@pbsds wrapped the binaries.

File: pkgs/by-name/xw/xwayland-run/package.nix [hunk 4]
```diff
@@ -0,0 +1,49 @@
+{ cage
+, fetchFromGitLab
+, gnome
+, lib
+, meson
+, ninja
+, python3
+, weston
+, xwayland
```
Comment (pbsds) [inline pkgs/by-name/xw/xwayland-run/package.nix:9]:
```suggestion
, xwayland
, withMutter ? false
, withCage ? false
```

File: pkgs/by-name/xw/xwayland-run/package.nix [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+}:
+
```
Comment (pbsds) [inline pkgs/by-name/xw/xwayland-run/package.nix:11]:
```suggestion

let
  compositors = [ weston ]
    ++ lib.optional withMutter gnome.mutter
    ++ lib.optional withCage cage
    ;
in
```

File: pkgs/by-name/xw/xwayland-run/package.nix [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  outputs = [ "out" "man" ];
+
+  nativeBuildInputs = [
+    meson
+    ninja
+  ];
+
+  postInstall = ''
+    wrapProgram $out/bin/wlheadless-run \
+      --prefix PATH : ${lib.makeBinPath [ weston cage gnome.mutter ]}
```
Comment (pbsds) [inline pkgs/by-name/xw/xwayland-run/package.nix:35]:
```suggestion
      --prefix PATH : ${lib.makeBinPath compositors}
```

File: pkgs/by-name/xw/xwayland-run/package.nix [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    wrapProgram $out/bin/xwayland-run \
+      --prefix PATH : ${lib.makeBinPath [ xwayland ]}
+    wrapProgram $out/bin/xwfb-run \
+      --prefix PATH : ${lib.makeBinPath [ weston cage gnome.mutter xwayland ]}
```
Comment (pbsds) [inline pkgs/by-name/xw/xwayland-run/package.nix:39]:
```suggestion
      --prefix PATH : ${lib.makeBinPath (compositors ++ [ xwayland ])}
```

File: pkgs/by-name/xw/xwayland-run/package.nix (hunk 4 above)
Comment (arthsmn) [reply pkgs/by-name/xw/xwayland-run/package.nix:9]: Done.

File: pkgs/by-name/xw/xwayland-run/package.nix (hunk 5 above)
Comment (arthsmn) [reply pkgs/by-name/xw/xwayland-run/package.nix]: Done.

File: pkgs/by-name/xw/xwayland-run/package.nix (hunk 6 above)
Comment (arthsmn) [reply pkgs/by-name/xw/xwayland-run/package.nix]: Done.

File: pkgs/by-name/xw/xwayland-run/package.nix (hunk 7 above)
Comment (arthsmn) [reply pkgs/by-name/xw/xwayland-run/package.nix]: Done.

**Comment by pbsds** [comment]
It seems `xwfb-run` hides this error, going by the strace, but i managed to invoke it such that we can see it:

```
[nix-shell:/dev/shm/nixpkgs-review/pr-273453-2]$ results/xwayland-run/bin/wlheadless-run -- results/xwayland-run/bin/xwayland-run -- env
...
Traceback (most recent call last):
  File "/nix/store/78l48rcbzmjnwiv29mj2pz3s6rbgyk73-xwayland-run-0.0.2/bin/..xwayland-run-wrapped-wrapped", line 70, in <module>
    if xwayland.create_xauth_entry() != 0:
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/78l48rcbzmjnwiv29mj2pz3s6rbgyk73-xwayland-run-0.0.2/lib/python3.11/site-packages/wlheadless/xwayland.py", line 108, in create_xauth_entry
    with subprocess.Popen(xauth_command_line, stdin = subprocess.PIPE) as proc:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/lib/python3.11/subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/nix/store/5k91mg4qjylxbfvrv748smfh51ppjq0g-python3-3.11.6/lib/python3.11/subprocess.py", line 1950, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'xauth'
```

It might need to be wrapped with `xorg.xauth`

**Comment by pbsds** [comment]
This should work headlessly, which can be emulated in a `--pure` nix shell ;)

**Comment by arthsmn** [comment]
Fixed.

**Review by pbsds — APPROVED** [review]
Result of `nixpkgs-review pr 273453` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>xwayland-run</li>
    <li>xwayland-run.man</li>
  </ul>
</details>

- [x] package path fits guidelines
- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] package builds on x86_64-linux
- [x] executables tested on x86_64-linux
- [x] `meta.description` is set and fits guidelines
- [x] `meta.license` fits upstream license
- [x] `meta.platforms` is set
- [x] `meta.maintainers` is set
- [x] build time only dependencies are declared in `nativeBuildInputs`
- [x] source is fetched using the appropriate function
- [x] the list of `phases` is not overridden
- [x] when a phase (like `installPhase`) is overridden it starts with `runHook preInstall` and ends with `runHook postInstall`.
- [x] patches have a comment describing either the upstream URL or a reason why the patch wasn't upstreamed
- [x] patches that are remotely available are fetched rather than vendored

**Comment by arthsmn** [comment]
@pbsds thanks for all your help, learned a lot here!

**Pull request merged**